README
NAME
    Net::Amazon::S3 - Use the Amazon S3 - Simple Storage Service

SYNOPSIS
      use Net::Amazon::S3;
      my $aws_access_key_id     = 'fill me in';
      my $aws_secret_access_key = 'fill me in too';

      my $s3 = Net::Amazon::S3->new(
          {   aws_access_key_id     => $aws_access_key_id,
              aws_secret_access_key => $aws_secret_access_key,
              retry                 => 1,
          }
      );

      # a bucket is a globally-unique directory
      # list all buckets that i own
      my $response = $s3->buckets;
      foreach my $bucket ( @{ $response->{buckets} } ) {
          print "You have a bucket: " . $bucket->bucket . "\n";
      }

      # create a new bucket
      my $bucketname = 'acmes_photo_backups';
      my $bucket = $s3->add_bucket( { bucket => $bucketname } )
          or die $s3->err . ": " . $s3->errstr;

      # or use an existing bucket
      $bucket = $s3->bucket($bucketname);

      # store a file in the bucket
      $bucket->add_key_filename( '1.JPG', 'DSC06256.JPG',
          { content_type => 'image/jpeg', },
      ) or die $s3->err . ": " . $s3->errstr;

      # store a value in the bucket
      $bucket->add_key( 'reminder.txt', 'this is where my photos are backed up' )
          or die $s3->err . ": " . $s3->errstr;

      # list files in the bucket
      $response = $bucket->list_all
          or die $s3->err . ": " . $s3->errstr;
      foreach my $key ( @{ $response->{keys} } ) {
          my $key_name = $key->{key};
          my $key_size = $key->{size};
          print "Bucket contains key '$key_name' of size $key_size\n";
      }

      # fetch file from the bucket
      $response = $bucket->get_key_filename( '1.JPG', 'GET', 'backup.jpg' )
          or die $s3->err . ": " . $s3->errstr;

      # fetch value from the bucket
      $response = $bucket->get_key('reminder.txt')
          or die $s3->err . ": " . $s3->errstr;
      print "reminder.txt:\n";
      print "  content length: " . $response->{content_length} . "\n";
      print "    content type: " . $response->{content_type} . "\n";
      print "            etag: " . $response->{content_type} . "\n";
      print "         content: " . $response->{value} . "\n";

      # delete keys
      $bucket->delete_key('reminder.txt') or die $s3->err . ": " . $s3->errstr;
      $bucket->delete_key('1.JPG')        or die $s3->err . ": " . $s3->errstr;

      # and finally delete the bucket
      $bucket->delete_bucket or die $s3->err . ": " . $s3->errstr;

DESCRIPTION
    This module provides a Perlish interface to Amazon S3. From the
    developer blurb: "Amazon S3 is storage for the Internet. It is designed
    to make web-scale computing easier for developers. Amazon S3 provides a
    simple web services interface that can be used to store and retrieve any
    amount of data, at any time, from anywhere on the web. It gives any
    developer access to the same highly scalable, reliable, fast,
    inexpensive data storage infrastructure that Amazon uses to run its own
    global network of web sites. The service aims to maximize benefits of
    scale and to pass those benefits on to developers".

    To find out more about S3, please visit: http://s3.amazonaws.com/

    To use this module you will need to sign up to Amazon Web Services and
    provide an "Access Key ID" and " Secret Access Key". If you use this
    module, you will incurr costs as specified by Amazon. Please check the
    costs. If you use this module with your Access Key ID and Secret Access
    Key you must be responsible for these costs.

    I highly recommend reading all about S3, but in a nutshell data is
    stored in values. Values are referenced by keys, and keys are stored in
    buckets. Bucket names are global.

    Note: This is the legacy interface, please check out
    Net::Amazon::S3::Client instead.

    Development of this code happens here:
    http://github.com/pfig/net-amazon-s3
    
    Homepage for the project (just started) is at
    http://pfig.github.com/net-amazon-s3/
    
METHODS
  new
    Create a new S3 client object. Takes some arguments:

    aws_access_key_id
        Use your Access Key ID as the value of the AWSAccessKeyId parameter
        in requests you send to Amazon Web Services (when required). Your
        Access Key ID identifies you as the party responsible for the
        request.

    aws_secret_access_key
        Since your Access Key ID is not encrypted in requests to AWS, it
        could be discovered and used by anyone. Services that are not free
        require you to provide additional information, a request signature,
        to verify that a request containing your unique Access Key ID could
        only have come from you.

        DO NOT INCLUDE THIS IN SCRIPTS OR APPLICATIONS YOU DISTRIBUTE.
        YOU'LL BE SORRY

    secure
        Set this to 1 if you want to use SSL-encrypted connections when
        talking to S3. Defaults to 0.

    timeout
        How many seconds should your script wait before bailing on a request
        to S3? Defaults to 30.

    retry
        If this library should retry upon errors. This option is
        recommended. This uses exponential backoff with retries after 1, 2,
        4, 8, 16, 32 seconds, as recommended by Amazon. Defaults to off.

  buckets
    Returns undef on error, else hashref of results

  add_bucket
    Takes a hashref:

    bucket
        The name of the bucket you want to add

    acl_short (optional)
        See the set_acl subroutine for documenation on the acl_short options

    location_constraint (option)
        Sets the location constraint of the new bucket. If left unspecified,
        the default S3 datacenter location will be used. Otherwise, you can
        set it to 'EU' for a European data center - note that costs are
        different.

    Returns 0 on failure, Net::Amazon::S3::Bucket object on success

  bucket BUCKET
    Takes a scalar argument, the name of the bucket you're creating

    Returns an (unverified) bucket object from an account. Does no network
    access.

  delete_bucket
    Takes either a Net::Amazon::S3::Bucket object or a hashref containing

    bucket
        The name of the bucket to remove

    Returns false (and fails) if the bucket isn't empty.

    Returns true if the bucket is successfully deleted.

  list_bucket
    List all keys in this bucket.

    Takes a hashref of arguments:

    MANDATORY

    bucket
        The name of the bucket you want to list keys on

    OPTIONAL

    prefix
        Restricts the response to only contain results that begin with the
        specified prefix. If you omit this optional argument, the value of
        prefix for your query will be the empty string. In other words, the
        results will be not be restricted by prefix.

    delimiter
        If this optional, Unicode string parameter is included with your
        request, then keys that contain the same string between the prefix
        and the first occurrence of the delimiter will be rolled up into a
        single result element in the CommonPrefixes collection. These
        rolled-up keys are not returned elsewhere in the response. For
        example, with prefix="USA/" and delimiter="/", the matching keys
        "USA/Oregon/Salem" and "USA/Oregon/Portland" would be summarized in
        the response as a single "USA/Oregon" element in the CommonPrefixes
        collection. If an otherwise matching key does not contain the
        delimiter after the prefix, it appears in the Contents collection.

        Each element in the CommonPrefixes collection counts as one against
        the MaxKeys limit. The rolled-up keys represented by each
        CommonPrefixes element do not. If the Delimiter parameter is not
        present in your request, keys in the result set will not be
        rolled-up and neither the CommonPrefixes collection nor the
        NextMarker element will be present in the response.

    max-keys
        This optional argument limits the number of results returned in
        response to your query. Amazon S3 will return no more than this
        number of results, but possibly less. Even if max-keys is not
        specified, Amazon S3 will limit the number of results in the
        response. Check the IsTruncated flag to see if your results are
        incomplete. If so, use the Marker parameter to request the next page
        of results. For the purpose of counting max-keys, a 'result' is
        either a key in the 'Contents' collection, or a delimited prefix in
        the 'CommonPrefixes' collection. So for delimiter requests, max-keys
        limits the total number of list results, not just the number of
        keys.

    marker
        This optional parameter enables pagination of large result sets.
        "marker" specifies where in the result set to resume listing. It
        restricts the response to only contain results that occur
        alphabetically after the value of marker. To retrieve the next page
        of results, use the last key from the current page of results as the
        marker in your next request.

        See also "next_marker", below.

        If "marker" is omitted,the first page of results is returned.

    Returns undef on error and a hashref of data on success:

    The hashref looks like this:

      {
            bucket          => $bucket_name,
            prefix          => $bucket_prefix, 
            common_prefixes => [$prefix1,$prefix2,...]
            marker          => $bucket_marker, 
            next_marker     => $bucket_next_available_marker,
            max_keys        => $bucket_max_keys,
            is_truncated    => $bucket_is_truncated_boolean
            keys            => [$key1,$key2,...]
       }

    Explanation of bits of that:

    common_prefixes
        If list_bucket was requested with a delimiter, common_prefixes will
        contain a list of prefixes matching that delimiter. Drill down into
        these prefixes by making another request with the prefix parameter.

    is_truncated
        B flag that indicates whether or not all results of your query were
        returned in this response. If your results were truncated, you can
        make a follow-up paginated request using the Marker parameter to
        retrieve the rest of the results.

    next_marker
        A convenience element, useful when paginating with delimiters. The
        value of "next_marker", if present, is the largest (alphabetically)
        of all key names and all CommonPrefixes prefixes in the response. If
        the "is_truncated" flag is set, request the next page of results by
        setting "marker" to the value of "next_marker". This element is only
        present in the response if the "delimiter" parameter was sent with
        the request.

    Each key is a hashref that looks like this:

         {
            key           => $key,
            last_modified => $last_mod_date,
            etag          => $etag, # An MD5 sum of the stored content.
            size          => $size, # Bytes
            storage_class => $storage_class # Doc?
            owner_id      => $owner_id,
            owner_displayname => $owner_name
        }

  list_bucket_all
    List all keys in this bucket without having to worry about 'marker'.
    This is a convenience method, but may make multiple requests to S3 under
    the hood.

    Takes the same arguments as list_bucket.

  add_key
    DEPRECATED. DO NOT USE

  get_key
    DEPRECATED. DO NOT USE

  head_key
    DEPRECATED. DO NOT USE

  delete_key
    DEPRECATED. DO NOT USE

LICENSE
    This module contains code modified from Amazon that contains the
    following notice:

      #  This software code is made available "AS IS" without warranties of any
      #  kind.  You may copy, display, modify and redistribute the software
      #  code either by itself or as incorporated into your code; provided that
      #  you do not remove any proprietary notices.  Your use of this software
      #  code is at your own risk and you waive any claim against Amazon
      #  Digital Services, Inc. or its affiliates with respect to your use of
      #  this software code. (c) 2006 Amazon Digital Services, Inc. or its
      #  affiliates.

TESTING
    Testing S3 is a tricky thing. Amazon wants to charge you a bit of money
    each time you use their service. And yes, testing counts as using.
    Because of this, the application's test suite skips anything approaching
    a real test unless you set these three environment variables:

    AMAZON_S3_EXPENSIVE_TESTS
        Doesn't matter what you set it to. Just has to be set

    AWS_ACCESS_KEY_ID
        Your AWS access key

    AWS_ACCESS_KEY_SECRET
        Your AWS sekkr1t passkey. Be forewarned that setting this
        environment variable on a shared system might leak that information
        to another user. Be careful.

AUTHOR
    Leon Brocard <acme@astray.com> and unknown Amazon Digital Services
    programmers.

    Brad Fitzpatrick <brad@danga.com> - return values, Bucket object
    
    Pedro Figueiredo <me@pedrofigueiredo.org> - since 0.54

SEE ALSO
    Net::Amazon::S3::Bucket



CHANGES
Revision history for Perl module Net::Amazon::S3:

{{$NEXT}}

0.59 Sat Mar 17 09:36:38 GMT 2013
    - Added support for security token authentication (patch by Miquel Ruiz)

0.58 Wed Nov 28 16:22:42 GMT 2012
    - Added support for multi-part upload and multiple objects deletion (patch
      by Robert B. Clarke)

0.57 Sun Sep 16 22:38:38 BST 2012
    - Added support for Content-disposition (initial patch by Michele Beltrame)
    - Added sync_up option to s3cl (patch by Leo Lapworth)
    - Dist::Zilla tidying

0.56 Sun Dec 18 22:22:32 GMT 2011
    - Spelling error (RT #69817, patch from Fabrizio Regalli)
    - Added use_ok tests (RT #72856, suggested by Gregor Herrmann)
    - Added missing use URI::Escape (RT #72857, patch from Gregor Herrmann)
    - Bumped dependency on MooseX::StrictConstructor to 0.16 (RT #73229,
      suggested by Andrew Main)
    - Enforce dependency versions (RT #73193, suggested by Andrew Main)

0.55 Sat Dec 10 00:55:29 GMT 2011
    - Replaced dependency on MooseX::Types::DateTimeX with
      MooseX::Types::DateTime::MoreCoercions (RT #72472).

0.54 Sat Mar 21 21:23:32 BST 2011
    - Fix for a naked qw() warning (patch by David Wheeler)
    - Fixed path issues (patch by Pavel Karoukin)
      *WARNING* THIS MIGHT BREAK EXISTING APPS *WARNING*
    - Author and development information changes

0.53 Tue Mar 30 15:24:19 BST 2010
    - fix authenticated urls to work with EU buckets (patch by Edmund
      von der Burg)
    - tiny POD fix (patch by Frank Wiegand)
    - add an exists method to Net::Amazon::S3::Client (suggested by
      David Golden)
    - fix max_keys when listing buckets (spotted by Andrew Bryan)
    - add content_encoding to Net::Amazon::S3::Object (suggested
      by Egor Korablev)
    - update s3cl: You need to use the module before you use it,
      added the mkbucket command, now you can run the help without
      your AWS secret key, add docs about the env variables you need
      to run s3cl (patches by Jesse Vincent)

0.52 Thu Jul  2 09:17:11 BST 2009
    - increase version prerequisites for some modules so that they
      are known to work

0.51 Tue May 19 08:31:59 BST 2009
    - use MooseX::Types::DateTimeX so that we work with latest Moose
      (noticed by Ted Zlatanov)

0.50 Wed Jan 21 10:42:00 GMT 2009
    - add support for an expires header when putting an object to
      Net::Amazon::S3::Client::Object
    - make all the classes immutable
    - add query_string_authentication_uri() to
      Net::Amazon::S3::Client::Object, suggested by Meng Wong

0.49 Tue Jan 13 09:04:42 GMT 2009
    - add support for listing a bucket with a prefix to
      Net::Amazon::S3::Client::Bucket

0.48 Thu Dec  4 09:24:23 GMT 2008
    - be slightly less strict about bucket names: they can contain uppercase
      letters, Amazon just doesn't recommend it (noticed by Simon Elliott,
      fixes Brackup)

0.47 Tue Dec  2 08:03:39 GMT 2008
    - fix listing with a prefix (spotted by Nobuo Danjou)

0.46 Mon Nov 24 08:53:18 GMT 2008
    - refactor request creation into Net::Amazon::S3::Request
      and many subclasses
    - move to Moose
    - add Net::Amazon::S3::Client and subclasses

0.45 Wed Aug 20 17:06:49 BST 2008
    - make add_key, head_key etc. return all the headers, not
      just the X-Amazon ones (patch by Andrew Hanenkamp)
    - require IO::File 1.14 (noticed by tsw)
    - remove DateTime::Format::Strptime prerequisite as it was not
      being used (noticed by Yen-Ming Lee)
    - do not try and parse non-XML errors (patch by lostlogic)
    - make it possible to store and delete the key "0"
      (patch by Joey Hess)
    - make it possible to store empty files (patch by BDOLAN)
    - add Copy support (patch by BDOLAN)
    - add s3cl for command-line access (patch by Leo Lapworth)

0.44 Thu Mar 27 08:35:59 GMT 2008
    - fix bug with storing files consisting of "0" (thanks to
      Martin Atkins)

0.43 Sat Mar  1 10:55:54 GMT 2008
     - add binmode() to support Windows (thanks to Gabriel Weinberg)

0.42 Thu Feb 28 06:39:59 GMT 2008
     - add exponential backoff upon temporary errors with the new
       retry option

0.41 Fri Nov 30 10:42:26 GMT 2007
     - fix the expensive tests (patch by BDOLAN)
     - added support for EU buckets (patch by BDOLAN)

0.40 Tue Oct 30 11:40:42 GMT 2007
     - fix for content length with empty keys by Mark A. Hershberger
     - get_key and get_key_filename now return content_length
     - rewrote synopsis
     - added support for common prefix (thanks to Andy Grundman)

0.39 Sun Aug 19 14:47:01 BST 2007
     - add add_key_filename and get_key_filename which send files
       directly from disk - good for large files (Jim Blomo)
     - support UTF8 keys (Jim Blomo)
     - switch back from Build.PL to Makefile.PL

0.38 Sun Mar  4 16:43:28 GMT 2007
     - use http_proxy and https_proxy environment variables for proxy
       settings (Ask Bjoern Hansen)
     - don't add the Authorization header if one is already specified
       when making a request - good for allowing caching to resources
       that are public. (Ask Bjoern Hansen)

0.37 Fri Oct 13 19:14:57 BST 2006
     - added support for ACLs (thanks to Gordon McCreight)

0.36 Sun Sep 10 16:30:39 BST 2006
     - remove extra warning

0.35 Sun Sep 10 16:25:44 BST 2006
     - added list_bucket_all to stop having to worrying about 'marker'

0.34 Sun Sep 10 07:27:06 BST 2006
     - added next marker and more docs from Jesse Vincent

0.33 Sat Aug 26 16:26:37 BST 2006
     - documentation and test cleanup from Jesse Vincent
     - use HTTP keep alive (patch by Scott Gifford)
     - remove ununused code in _make_request (patch by Scott Gifford)

0.32 Tue Apr 25 19:51:06 BST 2006
     - fix bug with listing buckets with parameters (thanks to karjala)

0.31 Tue Apr  4 21:15:02 BST 2006
     - many patches from Brad Fitzpatrick to make change the API, make
       return values sane, add err/errstr, make Bucket object
     - added a timeout option as suggested by Brad Dixon

lib/Net/Amazon/S3/Client/Object.pm
package Net::Amazon::S3::Client::Object;
use Moose 0.85;
use MooseX::StrictConstructor 0.16;
use DateTime::Format::HTTP;
use Digest::MD5 qw(md5 md5_hex);
use Digest::MD5::File qw(file_md5 file_md5_hex);
use File::stat;
use MIME::Base64;
use Moose::Util::TypeConstraints;
use MooseX::Types::DateTime::MoreCoercions 0.07 qw( DateTime );
use IO::File 1.14;

# ABSTRACT: An easy-to-use Amazon S3 client object

enum 'AclShort' =>
    qw(private public-read public-read-write authenticated-read);

has 'client' =>
    ( is => 'ro', isa => 'Net::Amazon::S3::Client', required => 1 );
has 'bucket' =>
    ( is => 'ro', isa => 'Net::Amazon::S3::Client::Bucket', required => 1 );
has 'key'  => ( is => 'ro', isa => 'Str',  required => 1 );
has 'etag' => ( is => 'ro', isa => 'Etag', required => 0 );
has 'size' => ( is => 'ro', isa => 'Int',  required => 0 );
has 'last_modified' =>
    ( is => 'ro', isa => DateTime, coerce => 1, required => 0 );
has 'expires' => ( is => 'rw', isa => DateTime, coerce => 1, required => 0 );
has 'acl_short' =>
    ( is => 'ro', isa => 'AclShort', required => 0, default => 'private' );
has 'content_type' => (
    is       => 'ro',
    isa      => 'Str',
    required => 0,
    default  => 'binary/octet-stream'
);
has 'content_disposition' => (
    is => 'ro',
    isa => 'Str',
    required => 0,
);
has 'content_encoding' => (
    is       => 'ro',
    isa      => 'Str',
    required => 0,
);

__PACKAGE__->meta->make_immutable;

sub exists {
    my $self = shift;

    my $http_request = Net::Amazon::S3::Request::GetObject->new(
        s3     => $self->client->s3,
        bucket => $self->bucket->name,
        key    => $self->key,
        method => 'HEAD',
    )->http_request;

    my $http_response = $self->client->_send_request_raw($http_request);
    return $http_response->code == 200 ? 1 : 0;
}

sub get {
    my $self = shift;

    my $http_request = Net::Amazon::S3::Request::GetObject->new(
        s3     => $self->client->s3,
        bucket => $self->bucket->name,
        key    => $self->key,
        method => 'GET',
    )->http_request;

    my $http_response = $self->client->_send_request($http_request);
    my $content       = $http_response->content;

    my $md5_hex = md5_hex($content);
    my $etag = $self->etag || $self->_etag($http_response);
    confess 'Corrupted download'
      if( !$self->_is_multipart_etag($etag) && $etag ne $md5_hex);

    return $content;
}

sub get_filename {
    my ($self, $filename) = @_;

    my $http_request = Net::Amazon::S3::Request::GetObject->new(
        s3     => $self->client->s3,
        bucket => $self->bucket->name,
        key    => $self->key,
        method => 'GET',
    )->http_request;

    my $http_response
        = $self->client->_send_request($http_request, $filename);

    my $md5_hex = file_md5_hex($filename);

    my $etag = $self->etag || $self->_etag($http_response);
    confess
      'Corrupted download' if(!$self->_is_multipart_etag($etag) && $etag ne $md5_hex);
}

sub put {
    my ( $self, $value ) = @_;
    my $md5        = md5($value);
    my $md5_hex    = unpack( 'H*', $md5 );
    my $md5_base64 = encode_base64($md5);
    chomp $md5_base64;

    my $conf = {
        'Content-MD5'    => $md5_base64,
        'Content-Length' => length $value,
        'Content-Type'   => $self->content_type,
    };

    if ( $self->expires ) {
        $conf->{Expires}
            = DateTime::Format::HTTP->format_datetime( $self->expires );
    }
    if ( $self->content_encoding ) {
        $conf->{'Content-Encoding'} = $self->content_encoding;
    }
    if ( $self->content_disposition ) {
        $conf->{'Content-Disposition'} = $self->content_disposition;
    }

    my $http_request = Net::Amazon::S3::Request::PutObject->new(
        s3        => $self->client->s3,
        bucket    => $self->bucket->name,
        key       => $self->key,
        value     => $value,
        headers   => $conf,
        acl_short => $self->acl_short,
    )->http_request;

    my $http_response = $self->client->_send_request($http_request);

    confess 'Error uploading' if $http_response->code != 200;

    my $etag = $self->_etag($http_response);

    confess 'Corrupted upload' if $etag ne $md5_hex;
}

sub put_filename {
    my ( $self, $filename ) = @_;

    my $md5_hex = $self->etag || file_md5_hex($filename);
    my $size = $self->size;
    unless ($size) {
        my $stat = stat($filename) || confess("No $filename: $!");
        $size = $stat->size;
    }

    my $md5 = pack( 'H*', $md5_hex );
    my $md5_base64 = encode_base64($md5);
    chomp $md5_base64;

    my $conf = {
        'Content-MD5'    => $md5_base64,
        'Content-Length' => $size,
        'Content-Type'   => $self->content_type,
    };

    if ( $self->expires ) {
        $conf->{Expires}
            = DateTime::Format::HTTP->format_datetime( $self->expires );
    }
    if ( $self->content_encoding ) {
        $conf->{'Content-Encoding'} = $self->content_encoding;
    }
    if ( $self->content_disposition ) {
        $conf->{'Content-Disposition'} = $self->content_disposition;
    }


    my $http_request = Net::Amazon::S3::Request::PutObject->new(
        s3        => $self->client->s3,
        bucket    => $self->bucket->name,
        key       => $self->key,
        value     => $self->_content_sub($filename),
        headers   => $conf,
        acl_short => $self->acl_short,
    )->http_request;

    my $http_response = $self->client->_send_request($http_request);

    confess 'Error uploading' . $http_response->as_string
        if $http_response->code != 200;

    confess 'Corrupted upload' if $self->_etag($http_response) ne $md5_hex;
}

sub delete {
    my $self = shift;

    my $http_request = Net::Amazon::S3::Request::DeleteObject->new(
        s3     => $self->client->s3,
        bucket => $self->bucket->name,
        key    => $self->key,
    )->http_request;

    $self->client->_send_request($http_request);
}

sub initiate_multipart_upload {
    my $self = shift;
    my $http_request = Net::Amazon::S3::Request::InitiateMultipartUpload->new(
        s3     => $self->client->s3,
        bucket => $self->bucket->name,
        key    => $self->key,
    )->http_request;
    my $xpc = $self->client->_send_request_xpc($http_request);
    my $upload_id = $xpc->findvalue('//s3:UploadId');
    confess "Couldn't get upload id from initiate_multipart_upload response XML"
      unless $upload_id;

    return $upload_id;
}

sub complete_multipart_upload {
    my $self = shift;

    my %args = ref($_[0]) ? %{$_[0]} : @_;

    #set default args
    $args{s3}       = $self->client->s3;
    $args{key}      = $self->key;
    $args{bucket}   = $self->bucket->name;

    my $http_request =
      Net::Amazon::S3::Request::CompleteMultipartUpload->new(%args)->http_request;
    return $self->client->_send_request($http_request);
}

sub put_part {
    my $self = shift;

    my %args = ref($_[0]) ? %{$_[0]} : @_;

    #set default args
    $args{s3}       = $self->client->s3;
    $args{key}      = $self->key;
    $args{bucket}   = $self->bucket->name;
    #work out content length header
    $args{headers}->{'Content-Length'} = length $args{value}
      if(defined $args{value});

    my $http_request =
      Net::Amazon::S3::Request::PutPart->new(%args)->http_request;
    return $self->client->_send_request($http_request);
}

sub list_parts {
    confess "Not implemented";
    # TODO - Net::Amazon::S3::Request:ListParts is implemented, but need to
    # define better interface at this level. Currently returns raw XML.
}

sub uri {
    my $self = shift;
    return Net::Amazon::S3::Request::GetObject->new(
        s3     => $self->client->s3,
        bucket => $self->bucket->name,
        key    => $self->key,
        method => 'GET',
    )->http_request->uri;
}

sub query_string_authentication_uri {
    my $self = shift;
    return Net::Amazon::S3::Request::GetObject->new(
        s3     => $self->client->s3,
        bucket => $self->bucket->name,
        key    => $self->key,
        method => 'GET',
    )->query_string_authentication_uri( $self->expires->epoch );
}

sub _content_sub {
    my $self      = shift;
    my $filename  = shift;
    my $stat      = stat($filename);
    my $remaining = $stat->size;
    my $blksize   = $stat->blksize || 4096;

    confess "$filename not a readable file with fixed size"
        unless -r $filename and ( -f _ || $remaining );
    my $fh = IO::File->new( $filename, 'r' )
        or confess "Could not open $filename: $!";
    $fh->binmode;

    return sub {
        my $buffer;

        # upon retries the file is closed and we must reopen it
        unless ( $fh->opened ) {
            $fh = IO::File->new( $filename, 'r' )
                or confess "Could not open $filename: $!";
            $fh->binmode;
            $remaining = $stat->size;
        }

        # warn "read remaining $remaining";
        unless ( my $read = $fh->read( $buffer, $blksize ) ) {

#                       warn "read $read buffer $buffer remaining $remaining";
            confess
                "Error while reading upload content $filename ($remaining remaining) $!"
                if $! and $remaining;

            # otherwise, we found EOF
            $fh->close
                or confess "close of upload content $filename failed: $!";
            $buffer ||= ''
                ;    # LWP expects an emptry string on finish, read returns 0
        }
        $remaining -= length($buffer);
        return $buffer;
    };
}

sub _etag {
    my ( $self, $http_response ) = @_;
    my $etag = $http_response->header('ETag');
    if ($etag) {
        $etag =~ s/^"//;
        $etag =~ s/"$//;
    }
    return $etag;
}

sub _is_multipart_etag {
    my ( $self, $etag ) = @_;
    return 1 if($etag =~ /\-\d+$/);
}

1;

__END__

=for test_synopsis
no strict 'vars'

=head1 SYNOPSIS

  # show the key
  print $object->key . "\n";

  # show the etag of an existing object (if fetched by listing
  # a bucket)
  print $object->etag . "\n";

  # show the size of an existing object (if fetched by listing
  # a bucket)
  print $object->size . "\n";

  # to create a new object
  my $object = $bucket->object( key => 'this is the key' );
  $object->put('this is the value');

  # to get the vaue of an object
  my $value = $object->get;

  # to see if an object exists
  if ($object->exists) { ... }

  # to delete an object
  $object->delete;

  # to create a new object which is publically-accessible with a
  # content-type of text/plain which expires on 2010-01-02
  my $object = $bucket->object(
    key          => 'this is the public key',
    acl_short    => 'public-read',
    content_type => 'text/plain',
    expires      => '2010-01-02',
  );
  $object->put('this is the public value');

  # return the URI of a publically-accessible object
  my $uri = $object->uri;

  # upload a file
  my $object = $bucket->object(
    key          => 'images/my_hat.jpg',
    content_type => 'image/jpeg',
  );
  $object->put_filename('hat.jpg');

  # upload a file if you already know its md5_hex and size
  my $object = $bucket->object(
    key          => 'images/my_hat.jpg',
    content_type => 'image/jpeg',
    etag         => $md5_hex,
    size         => $size,
  );
  $object->put_filename('hat.jpg');

  # download the value of the object into a file
  my $object = $bucket->object( key => 'images/my_hat.jpg' );
  $object->get_filename('hat_backup.jpg');

  # use query string authentication
  my $object = $bucket->object(
    key          => 'images/my_hat.jpg',
    expires      => '2009-03-01',
  );
  my $uri = $object->query_string_authentication_uri();

=head1 DESCRIPTION

This module represents objects in buckets.

=head1 METHODS

=head2 etag

  # show the etag of an existing object (if fetched by listing
  # a bucket)
  print $object->etag . "\n";

=head2 delete

  # to delete an object
  $object->delete;

=head2 exists

  # to see if an object exists
  if ($object->exists) { ... }

=head2 get

  # to get the vaue of an object
  my $value = $object->get;

=head2 get_filename

  # download the value of the object into a file
  my $object = $bucket->object( key => 'images/my_hat.jpg' );
  $object->get_filename('hat_backup.jpg');

=head2 key

  # show the key
  print $object->key . "\n";

=head2 put

  # to create a new object
  my $object = $bucket->object( key => 'this is the key' );
  $object->put('this is the value');

  # to create a new object which is publically-accessible with a
  # content-type of text/plain
  my $object = $bucket->object(
    key          => 'this is the public key',
    acl_short    => 'public-read',
    content_type => 'text/plain',
  );
  $object->put('this is the public value');

You may also set Content-Encoding using content_encoding, and
Content-Disposition using content_disposition.

=head2 put_filename

  # upload a file
  my $object = $bucket->object(
    key          => 'images/my_hat.jpg',
    content_type => 'image/jpeg',
  );
  $object->put_filename('hat.jpg');

  # upload a file if you already know its md5_hex and size
  my $object = $bucket->object(
    key          => 'images/my_hat.jpg',
    content_type => 'image/jpeg',
    etag         => $md5_hex,
    size         => $size,
  );
  $object->put_filename('hat.jpg');

You may also set Content-Encoding using content_encoding, and
Content-Disposition using content_disposition.

=head2 query_string_authentication_uri

  # use query string authentication
  my $object = $bucket->object(
    key          => 'images/my_hat.jpg',
    expires      => '2009-03-01',
  );
  my $uri = $object->query_string_authentication_uri();

=head2 size

  # show the size of an existing object (if fetched by listing
  # a bucket)
  print $object->size . "\n";

=head2 uri

  # return the URI of a publically-accessible object
  my $uri = $object->uri;

=head2 initiate_multipart_upload

  #initiate a new multipart upload for this object
  my $object = $bucket->object(
    key         => 'massive_video.avi'
  );
  my $upload_id = $object->initiate_multipart_upload;

=head2 put_part

  #add a part to a multipart upload
  my $put_part_response = $object->put_part(
     upload_id      => $upload_id,
     part_number    => 1,
     value          => $chunk_content,
  );
  my $part_etag = $put_part_response->header('ETag')

  Returns an L<HTTP::Response> object. It is necessary to keep the ETags for
  each part, as these are required to complete the upload.

=head2 complete_multipart_upload

  #complete a multipart upload
  $object->complete_multipart_upload(
    upload_id       => $upload_id,
    etags           => [$etag_1, $etag_2],
    part_numbers    => [$part_number_1, $part_number2],
  );

  The etag and part_numbers parameters are ordered lists specifying the part
  numbers and ETags for each individual part of the multipart upload.
